### PR TITLE
Fix Sorbet ignore matching

### PR DESCRIPTION
**Pull request by paracycle**
### Motivation

I pushed a change to cache bundler installed gems on `master` which ended up breaking tests. The change installs gems into the `vendor/bundle` folder. Even though `/vendor` is an `--ignore` flag in `sorbet/config`, our tests were somehow still being affected by the existence of the `vendor` folder.

I investigated the cause to Sorbet ignore matching not exactly working exactly like a regular expression match. Instead, it tries to match whole path segments unless the ignore path starts with a `/` character in which case it tries to match the prefix of relative paths.

### Implementation

The new implementation first tries to get the relative path of each found file to the working path. All matches are done against this relative path. For each ignore string, if it starts with a `/`, the first segment of the relative path. Otherwise, we expect it to match a whole part of the relative path.

### Tests

I haven't added any extra tests, but all the existing ones pass.

### CODE REVIEW

File: lib/tapioca/compilers/requires_compiler.rb [hunk 1]
```diff
@@ -49,13 +50,40 @@ def collect_requires(file_path)
...
+          file_path_part_sequences = [file_path_part_sequences.first].to_enum if ignore.start_with?("/")
+
+          # we need to match whole segments
+          file_path_part_sequences.include?(ignore_parts)
         end
       end
 
+      sig { params(path: Pathname).returns(T::Array[String]) }
+      def path_parts(path)
+        T.unsafe(path).descend.map { |part| part.basename.to_s }
```
Comment (rafaelfranca) [inline lib/tapioca/compilers/requires_compiler.rb:84]: Why the unsafe here. Isn’t descend in the Pathname class?

**Review by rafaelfranca — APPROVED** [review]

File: lib/tapioca/compilers/requires_compiler.rb (hunk 1 above)
Comment (paracycle) [reply lib/tapioca/compilers/requires_compiler.rb:84]: Sorry I forgot to note that. It is because of this: https://github.com/sorbet/sorbet/pull/3617

Comment (paracycle) [reply lib/tapioca/compilers/requires_compiler.rb:84]: Basically, the block argument should be optional but isn't

File: lib/tapioca/compilers/requires_compiler.rb [hunk 2]
```diff
@@ -49,13 +50,40 @@ def collect_requires(file_path)
         end.compact
       end
 
-      sig { params(config: Spoom::Sorbet::Config, file: String).returns(T::Boolean) }
-      def file_ignored_by_sorbet?(config, file)
-        config.ignore.any? do |path|
-          Regexp.new(Regexp.escape(path)) =~ file
+      sig { params(config: Spoom::Sorbet::Config, file_path: Pathname).returns(T::Boolean) }
+      def file_ignored_by_sorbet?(config, file_path)
```
Comment (Morriar) [inline lib/tapioca/compilers/requires_compiler.rb:54]: Should we move this service to `Spoom`? Seems useful.

**Review by Morriar — APPROVED** [review]

File: lib/tapioca/compilers/requires_compiler.rb (hunk 2 above)
Comment (paracycle) [reply lib/tapioca/compilers/requires_compiler.rb:54]: Yeah, I was thinking the same thing. Let's do it in a future refactor, and we can add tests around it as well.

**Pull request merged**